lib/client/keyagent.go

```diff
@@ -258,77 +258,99 @@ func (a *LocalKeyAgent) UserRefusedHosts() bool {
-
-			bytes := make([]byte, 12)
-			os.Stdin.Read(bytes)
-			userAnswer = strings.TrimSpace(strings.ToLower(string(bytes)))
-		}
-		if !strings.HasPrefix(userAnswer, "y") {
-			return trace.AccessDenied("untrusted host %v", host)
-		}
-		// success
-		return nil
+// CheckHostSignature checks if the given host key was signed by a Teleport
+// certificate authority (CA) or a host certificate the user has seen before.
+func (a *LocalKeyAgent) CheckHostSignature(addr string, remote net.Addr, key ssh.PublicKey) error {
+	certChecker := ssh.CertChecker{
+		IsHostAuthority: a.checkHostCertificate,
+		HostKeyFallback: a.checkHostKey,
 	}
-	// overwritten host prompt func? (probably for tests)
-	if a.hostPromptFunc != nil {
-		hostPromptFunc = a.hostPromptFunc
-	}
-	cert, ok := key.(*ssh.Certificate)
-	if !ok {
-		// not a signed cert? perhaps we're given a host public key (happens when the host is running
-		// sshd instead of teleport daemon
-		keys, _ := a.keyStore.GetKnownHostKeys(host)
-		if len(keys) > 0 && sshutils.KeysEqual(key, keys[0]) {
-			a.log.Debugf("Verified host %s", host)
-			return nil
-		}
-		// ask user:
-		if err := hostPromptFunc(host, key); err != nil {
-			a.noHosts[host] = true
-			return trace.Wrap(err)
-		}
-		// remember the host key (put it into 'known_hosts')
-		if err := a.keyStore.AddKnownHostKeys(host, []ssh.PublicKey{key}); err != nil {
-			a.log.Warnf("Error saving the host key: %v", err)
-		}
-		return nil
+	err := certChecker.CheckHostKey(addr, remote, key)
+	if err != nil {
+		a.log.Debugf("Host validation failed: %v.", err)
+		return trace.Wrap(err)
 	}
-	key = cert.SignatureKey
-	// we are given a certificate. see if it was signed by any of the known_host keys:
+	a.log.Debugf("Validated host %v.", addr)
+	return nil
+}
+
+// checkHostCertificate validates a host certificate. First checks the
+// ~/.tsh/known_hosts cache and if not found, prompts the user to accept
+// or reject.
+func (a *LocalKeyAgent) checkHostCertificate(key ssh.PublicKey, addr string) bool {
+	// Check the local cache (where all Teleport CAs are placed upon login) to
+	// see if any of them match.
 	keys, err := a.keyStore.GetKnownHostKeys("")
 	if err != nil {
-		a.log.Error(err)
-		return trace.Wrap(err)
+		a.log.Errorf("Unable to fetch certificate authorities: %v.", err)
+		return false
 	}
-	a.log.Debugf("Got %d known hosts", len(keys))
 	for i := range keys {
-		if sshutils.KeysEqual(cert.SignatureKey, keys[i]) {
-			a.log.Debugf("Verified host %s", host)
-			return nil
+		if sshutils.KeysEqual(key, keys[i]) {
+			return true
 		}
 	}
-	// final step: if we have not seen the host key/cert before, lets ask the user if
-	// he trusts it, and add to the known_hosts if he says "yes"
-	if err = hostPromptFunc(host, key); err != nil {
-		// he said "no"
-		a.noHosts[host] = true
+
+	// If this certificate was not seen before, prompt the user essentially
+	// treating it like a key.
+	err = a.checkHostKey(addr, nil, key)
```

Comment:
isn't this redundant with `HostKeyFallback` in place?

Reply:
This prompts the user to trust a host certificate, in `HostKeyFallback` the user is prompted to check a host key.